spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/integration/outbound/BigQueryFileMessageHandler.java

```diff
@@ -36,30 +36,97 @@
+
+	private FormatOptions formatOptions;
+
 	private boolean autoDetectSchema = true;
 
 	private WriteDisposition writeDisposition = WriteDisposition.WRITE_APPEND;
 
 	public BigQueryFileMessageHandler(BigQuery bigQuery) {
+		Assert.notNull(bigQuery, "BigQuery client must not be null.");
 		this.bigQuery = bigQuery;
 	}
 
+	/**
+	 * Sets the default BigQuery dataset name to use if it is omitted from message headers.
+	 * @param datasetName name of the BigQuery dataset
+	 */
+	public void setDatasetName(String datasetName) {
+		this.datasetName = datasetName;
+	}
+
+	/**
+	 * Sets the default BigQuery table name to use if it is omitted from message headers.
+	 * @param tableName name of the BigQuery dataset
+	 */
+	public void setTableName(String tableName) {
+		this.tableName = tableName;
+	}
+
+	/**
+	 * Sets the default {@link FormatOptions} to use if it is omitted from message headers.
+	 * The {@link FormatOptions} describe the type/format of data files being loaded.
+	 * @param formatOptions the format of the data file being loaded
+	 */
+	public void setFormatOptions(FormatOptions formatOptions) {
+		this.formatOptions = formatOptions;
+	}
+
+	/**
+	 * Sets the {@link WriteDisposition} which specifies how data should be inserted into
+	 * BigQuery tables.
+	 * @param writeDisposition whether data should be appended or truncated to the BigQuery table
+	 */
+	public void setWriteDisposition(WriteDisposition writeDisposition) {
+		this.writeDisposition = writeDisposition;
+	}
+
+	/**
+	 * Sets whether BigQuery should attempt to autodetect the schema of the data when loading
+	 * data into an empty table for the first time.
+	 * @param autoDetectSchema whether data schema should be autodetected from the structure of
+	 * 		the data
+	 */
+	public void setAutoDetectSchema(boolean autoDetectSchema) {
+		this.autoDetectSchema = autoDetectSchema;
+	}
+
 	@Override
 	protected Object handleRequestMessage(Message<?> message) {
 		try {
-			String datasetName = message.getHeaders().get(BigQuerySpringMessageHeaders.DATASET_NAME, String.class);
-			String tableName = message.getHeaders().get(BigQuerySpringMessageHeaders.TABLE_NAME, String.class);
-			FormatOptions formatOptions = message.getHeaders().get(
-					BigQuerySpringMessageHeaders.FORMAT_OPTIONS, FormatOptions.class);
+			String datasetName =
+					message.getHeaders().containsKey(BigQuerySpringMessageHeaders.DATASET_NAME)
+							? message.getHeaders().get(BigQuerySpringMessageHeaders.DATASET_NAME, String.class)
+							: this.datasetName;
+
+			String tableName =
+					message.getHeaders().containsKey(BigQuerySpringMessageHeaders.TABLE_NAME)
+							? message.getHeaders().get(BigQuerySpringMessageHeaders.TABLE_NAME, String.class)
+							: this.tableName;
+
+			FormatOptions formatOptions =
+					message.getHeaders().containsKey(BigQuerySpringMessageHeaders.FORMAT_OPTIONS)
+							? message.getHeaders().get(
+									BigQuerySpringMessageHeaders.FORMAT_OPTIONS, FormatOptions.class)
```

Comment:
Is really `FormatOptions` a thing to be carried in headers?
Is it really so message-specific?
Is it JSON-, Serialization-compatible?

Reply:
Is really FormatOptions a thing to be carried in headers?
- Hmm, I am considering the use-case where somebody might have a couple different file formats they wish to load to the database; imagine having both CSV and JSON. In that case, I thought it would be reasonable to let the user specify the file format in the message header.

Is it really so message-specific?
- Based on the above reasoning if correct then it would be message specific.

Is it JSON-, Serialization-compatible?
- It [implements Serializable](https://github.com/googleapis/google-cloud-java/blob/master/google-cloud-clients/google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/FormatOptions.java), however unsure about the JSON thing.

Reply:
Well, I think that existing ctor should make it compatible with JSON (de)serialization even if it is a package protected.
So, this is fine with me so far.